File: external_contracts/build_contracts.py
import json


class GlobalContracts:
    def __init__(self, web3, path):
        self.w3 = web3
        self.path = path
        self._build_contracts()

    def _build_contracts(self):
        self.contracts = dict()
        with open(self.path) as file:
            contracts = json.load(file)
        for protocol in contracts.keys():
            for contract in contracts[protocol].keys():
                address = contracts[protocol][contract]['address']
                abi = contracts[protocol][contract]['abi']
                code = f"self.{protocol}_{contract} = self.w3.eth.contract(address='{address}', abi={abi})"
                exec(code)

File: src/external_contracts/build_contracts.rs
use std::fs::File;
use std::io::BufReader;
use std::collections::HashMap;
use std::error::Error;
use serde_json::Value;
use web3::contract::{Contract, ContractError};
use web3::eth::Eth;
use web3::types::Address;

struct GlobalContracts {
    web3: Eth,
    path: String,
    contracts: HashMap<String, Contract<Eth>>,
}

impl GlobalContracts {
    fn new(web3: Eth, path: String) -> Result<Self, Box<dyn Error>> {
        let mut contracts = HashMap::new();
        let file = File::open(&path)?;
        let reader = BufReader::new(file);
        let contracts_json: Value = serde_json::from_reader(reader)?;

        for (protocol, contract_map) in contracts_json.as_object().unwrap().iter() {
            for (contract, data) in contract_map.as_object().unwrap().iter() {
                let address: Address = data["address"].as_str().unwrap().parse()?;
                let abi = data["abi"].clone();
                let key = format!("{}_{}", protocol, contract);
                let contract_instance = Contract::from_json(web3.clone(), address, abi.to_string().as_bytes())?;
                contracts.insert(key, contract_instance);
            }
        }

        Ok(Self {
            web3,
            path,
            contracts,
        })
    }
}
